project/frontend/src/pages/Header.js

```diff
@@ -1,24 +1,105 @@
 import './Header.css';
 
-import React from 'react';
+import React, { useState, useEffect } from 'react';
+
+import Modal from '../components/Modal.js';
+import Loading from '../components/Loading.js';
 
 function Header(props) {
-  return (
-    <div id='header'>
-      {props.isUserLoggedIn ? (
-        <div>Pic</div>
-      ) : (
-        [
-          <button key='1' className='header-button'>
-            Create Account
-          </button>,
-          <button key='2' className='header-button'>
-            Log in
-          </button>,
-        ]
-      )}
-    </div>
-  );
+  let [GoogleAuth, setGoogleAuth] = useState(undefined);
+  let [currentUser, setCurrentUser] = useState(undefined);
+  let [signedIn, setSignedIn] = useState(false);
```

Comment:
Do we need to have both currentUser and signedIn? Can we just use currentUser.isSignedIn to determine whether or not the user is signed in?

Reply:
No I don't think we need both. The problem is that React re-renders a component when you update its state through the setters returned by `useState(...)` (e.g. `setSignedIn`, `setCurrentUser`), but will not actually re-paint a component if you update the state with an equal value. So when a user signs in, `setCurrentUser(...)` is called with the new user object, but the Google API uses the same object in memory and just overwrites the internal pointers, so react doesn't think the object has changed and then doesn't re-paint the component after re-rendering.

React uses `Object.is(...)` to test state equality which doesn't account for change to internal values inside an object, just checks the memory location. I tried cloning the user object when `setCurrentUser(...)` is called but this did not work, cloning objects in JavaScript turns out to be a lot more difficult that I would've expected. I also tried adding a new property with a random value to the user object every time `setCurrentUser(...)` is called, but of course again didn't work since the memory location is being compared, not the properties of the object.

My workaround to this problem is defining the `signedIn` state which is just a boolean instead of an object. Now when the user changes both `setCurrentUser(...)` and `setSignedIn(...)` are called with new values and `Object.is(...)` recognizes `false` and `true` as not being equal so React fully re-renders and re-paints the component.

Reply:
This problem is still a problem after the refactor using a global context state and Authentication component. An optimal solution would be to have React recognize that the `currentUser` object has changed but I'm not sure how to do this.